From bc669514eeb98bc8ca8c0d2e9b713d87ddb892cc Mon Sep 17 00:00:00 2001
From: Gauthier Haderer <ghaderer@wyplay.com>
Date: Thu, 11 May 2017 10:01:28 +0000
Subject: [PATCH] Propagate KEY_MENU, KEY_INFO, KEY_CHANNELUP and
 KEY_CHANNELDOWN to JS.

At the JS level, we get the following values:
 - KEY_MENU: keyCode=0, key=MENU, code=MENU
 - KEY_INFO: keyCode=0, key=INFO, code=INFO
 - KEY_CHANNELUP: keyCode=0, key=ChannelUp, code=ChannelUp
 - KEY_CHANNELDOWN: keyCode=0, key=ChannelDown, code=ChannelDown

Note that the MENU key is not part of the W3C "UI Events KeyboardEvent
key Values" and "UI Events KeyboardEvent code Values" standards. This is
an extension as I could not find a key with the appropriate meaning.
---
 ui/events/keycodes/dom/dom_key_data.inc           | 2 ++
 ui/events/keycodes/dom/keycode_converter_data.inc | 9 +++++----
 ui/events/keycodes/dom_us_layout_data.h           | 2 +-
 3 files changed, 8 insertions(+), 5 deletions(-)

diff --git ui/events/keycodes/dom/dom_key_data.inc ui/events/keycodes/dom/dom_key_data.inc
index 507d97e..ed318bf 100644
--- ui/events/keycodes/dom/dom_key_data.inc
+++ ui/events/keycodes/dom/dom_key_data.inc
@@ -160,6 +160,8 @@ DOM_KEY_MAP_DECLARATION {
   DOM_KEY_MAP("Select",               SELECT,                   0x050C),
   DOM_KEY_MAP("ZoomIn",               ZOOM_IN,                  0x050D),
   DOM_KEY_MAP("ZoomOut",              ZOOM_OUT,                 0x050E),
+  //          Non-standard extension
+  DOM_KEY_MAP("Menu",                 MENU,                     0x05FF),
 
   // ========================================================
   // Device Keys
diff --git ui/events/keycodes/dom/keycode_converter_data.inc ui/events/keycodes/dom/keycode_converter_data.inc
index a62d390..4287c00 100644
--- ui/events/keycodes/dom/keycode_converter_data.inc
+++ ui/events/keycodes/dom/keycode_converter_data.inc
@@ -277,7 +277,8 @@ USB_KEYMAP_DECLARATION {
   // USB#0x070075 Help -- see note M1 at top.
   USB_KEYMAP(0x070075, 0x008a, 0x0092, 0xe03b, 0xffff, "Help", HELP),
   // USB#0x070076 Keyboard Menu -- see note L2 at top.
-  //USB_KEYMAP(0x070076, 0x0000, 0x0000, 0x0000, 0xffff, NULL, MENU), // Menu
+  // Non-standard key "Menu"
+  USB_KEYMAP(0x070076, 0x008b, 0x0093, 0x0000, 0xffff, "Menu", MENU), // Menu
   USB_KEYMAP(0x070077, 0x0084, 0x008c, 0x0000, 0xffff, "Select", SELECT), // Select
 
   //USB_KEYMAP(0x070078, 0x0080, 0x0088, 0x0000, 0xffff, NULL, STOP), // Stop
@@ -457,7 +458,7 @@ USB_KEYMAP_DECLARATION {
   // KEY_BRIGHTNESS* added in Linux 3.16
   // http://www.usb.org/developers/hidpage/HUTRR41.pdf
   //            USB     evdev    XKB     Win     Mac   Code
-  USB_KEYMAP(0x0c0060, 0x0166, 0x016e, 0x0000, 0xffff, NULL, INFO),
+  USB_KEYMAP(0x0c0060, 0x0166, 0x016e, 0x0000, 0xffff, "Info", INFO),
   USB_KEYMAP(0x0c0061, 0x0172, 0x017a, 0x0000, 0xffff, NULL, CLOSED_CAPTION_TOGGLE),
   USB_KEYMAP(0x0c006f, 0x00e1, 0x00e9, 0x0000, 0xffff, "BrightnessUp", BRIGHTNESS_UP),
   USB_KEYMAP(0x0c0070, 0x00e0, 0x00e8, 0x0000, 0xffff, "BrightnessDown",
@@ -470,8 +471,8 @@ USB_KEYMAP_DECLARATION {
   USB_KEYMAP(0x0c008c, 0x00a9, 0x00b1, 0x0000, 0xffff, NULL, LAUNCH_PHONE),
   USB_KEYMAP(0x0c008d, 0x016a, 0x0172, 0x0000, 0xffff, NULL, PROGRAM_GUIDE),
   USB_KEYMAP(0x0c0094, 0x00ae, 0x00b6, 0x0000, 0xffff, NULL, EXIT),
-  USB_KEYMAP(0x0c009c, 0x019a, 0x01a2, 0x0000, 0xffff, NULL, CHANNEL_UP),
-  USB_KEYMAP(0x0c009d, 0x019b, 0x01a3, 0x0000, 0xffff, NULL, CHANNEL_DOWN),
+  USB_KEYMAP(0x0c009c, 0x0192, 0x019a, 0x0000, 0xffff, "ChannelUp", CHANNEL_UP),
+  USB_KEYMAP(0x0c009d, 0x0193, 0x019b, 0x0000, 0xffff, "ChannelDown", CHANNEL_DOWN),
 
   //              USB     evdev    XKB     Win     Mac
   USB_KEYMAP(0x0c00b0, 0x00cf, 0x00d7, 0x0000, 0xffff, "MediaPlay", MEDIA_PLAY),
diff --git ui/events/keycodes/dom_us_layout_data.h ui/events/keycodes/dom_us_layout_data.h
index 4cf29e9..c9ced3d 100644
--- ui/events/keycodes/dom_us_layout_data.h
+++ ui/events/keycodes/dom_us_layout_data.h
@@ -205,7 +205,7 @@ const struct NonPrintableCodeEntry {
     {DomCode::MEDIA_STOP, DomKey::MEDIA_STOP},
     {DomCode::MEDIA_TRACK_NEXT, DomKey::MEDIA_TRACK_NEXT},
     {DomCode::MEDIA_TRACK_PREVIOUS, DomKey::MEDIA_TRACK_PREVIOUS},
-    // {DomCode::MENU, DomKey::_}
+    {DomCode::MENU, DomKey::MENU},
     {DomCode::NEW, DomKey::NEW},
     {DomCode::NON_CONVERT, DomKey::NON_CONVERT},
     {DomCode::NUM_LOCK, DomKey::NUM_LOCK},
-- 
1.8.5.3

